playwright/utils/MockManagerProcedures.ts

```diff
@@ -0,0 +1,113 @@
+import { Page, Route } from "@playwright/test";
+import { encode, decode } from "@msgpack/msgpack";
+
+type MockManagerProcedure = {
+  handler: MockManagerProcedureHandler;
+  config: MockManagerProcedureConfig;
+};
+
+type MockManagerProcedureHandler = (args: {
+  data?: unknown;
+  args: unknown[];
+}) => unknown | Promise<unknown>;
+
+type MockManagerProcedureConfig = {
+  execute?: boolean;
+  times?: number;
+};
+
+const FORCE_INIT = Symbol();
+
+export class MockManagerProcedures {
+  static async init(page: Page) {
+    const mockManagerProcedures = new MockManagerProcedures(FORCE_INIT);
+
+    await page.route(
+      "*/**/_manager",
+      mockManagerProcedures.#handleRoute.bind(mockManagerProcedures),
+    );
+
+    return mockManagerProcedures;
+  }
+
+  procedures = new Map<string, MockManagerProcedure[]>();
+
+  /**
+   * @depreacted Use `MockManagerProcedures.init` instead.
+   */
+  constructor(DO_NOT_USE_CONSTRUCTOR: typeof FORCE_INIT) {
+    if (DO_NOT_USE_CONSTRUCTOR !== FORCE_INIT) {
+      throw new Error(
+        "`new MockManagerProcedures()` is forbidden. Use `await MockManagerProcedures.init(page)` instead.",
+      );
+    }
+  }
+
+  mock(
+    path: string,
```

Comment:
I think I would keep the object, you may not need the handler for example and so positioning argument can be painful, WDYT?

Reply:
Requiring a handler feels more explicit to me. By defining a `() => undefined` handler, the test tells readers that calling the manager procedure will return undefined. Without a handler, that functionality is implicit.

I typically prefer objects as well since they document argument names and allow for easier argument omission. In this case, I figured the `path` and `handler` arguments are always needed, allowing using to create a more compact API.

The positional arguments also make reading the call more straightforward in my opinion.

```ts
procedures.mock("project.fetchActiveEnvironment", () => ({ environment }));

// Or a more compact version without a function when
// a function isn't needed (not implemented yet):
procedures.mock("project.fetchActiveEnvironment", { environment });
```

Compared to:

```ts
procedures.mock({
  path: "project.fetchActiveEnvironment",
  handler: () => ({ environment }),
});

// Or a more compact version without a function when
// a function isn't needed (not implemented yet):
procedures.mock({
  path: "project.fetchActiveEnvironment",
  // The `handler` name no longer makes sense since it isn't a function.
  handler: { environment },
});
```

"Handler" is more accurate than "data" since the function may throw an error.

I don't mind either option, but I prefer the non-object version. If we feel strongly about making the handler optional, then I agree, an object is better.

Reply:
You convince me! 🙂 Let's go your way